### PR TITLE
Preserve DevTools package contract in release and PR snapshots

### LINKED ISSUE #1454: Preserve DevTools peer dependency contract during artifact repack
## Problem

`scripts/src/commands/devtools-artifact.ts repack` currently replaces the artifact package dependency shape: it moves Vite and LiveStore core packages into runtime `dependencies` and widens Vite to `*`. This discards the artifact-owned Vite peer range and makes consumers install a second runtime dependency surface that should be supplied by the host application.

## Desired behavior

- Preserve every peer dependency range declared by the verified artifact.
- Pin `@livestore/adapter-web` and `@livestore/utils` as exact peers matching the target LiveStore snapshot or release version.
- Keep only `@parcel/watcher` in runtime `dependencies`, because its native loader remains external to the bundle.
- Fail the repack if the resulting core peer contract is not exact or the required watcher dependency is absent.
- Cover snapshot and release targets with focused manifest-shape regression tests, including the Vite range `^7.3.1 || ^8.0.16`.

No artifact publication is part of this issue.

<details>
<summary>Posted on behalf of @schickling</summary>

| field | value |
| --- | --- |
| `agent_name` | ☁️ co3-billow |
| `agent_session_id` | 7f3b9666-034f-4e9e-890b-491957f7f75b |
| `agent_tool` | Codex CLI |
| `agent_tool_version` | 0.144.1 |
| `agent_runtime` | Codex CLI 0.144.1 |
| `agent_model` | unknown |
| `runtime_profile` | /nix/store/1n76sy6i9y3ki3k4w04jksqvygw67sj0-coding-agent-runtime-profile/share/coding-agents/profile.json |
| `skills_manifest` | /nix/store/jn6r0r7r59x3a525jch4pwzwykszqp60-agent-skills-corpus/share/agent-skills/manifest.json |
| `worktree` | livestore-devtools-artifact-peer-deps/schickling-assistant/fix/devtools-artifact-peer-deps |
| `machine` | dev3 |
| `tooling_profile` | dotfiles@4eac376 |
</details>

### PR DESCRIPTION
**Pull request by schickling-assistant**
## Problem

The DevTools artifact repacker discarded the verified artifact's peer contract: it moved Vite and LiveStore core packages into runtime dependencies and widened Vite to `*`. Separately, trusted PR snapshots packed only the 14 workspace-owned packages, so `@livestore/devtools-vite` could not be installed at the exact snapshot version used by downstream integration tests.

## Goal

Preserve the artifact-owned host compatibility contract, exact-pin its LiveStore peers, and publish DevTools as part of the same immutable 15-package trusted PR snapshot cohort.

## Decisions

- Preserve every artifact peer range except `@livestore/adapter-web` and `@livestore/utils`; bind those two peers to the exact target version.
- Retain only the externalized native `@parcel/watcher` package as a runtime dependency and fail closed if it is absent.
- Keep workspace release discovery unchanged at 14 packages. Model DevTools explicitly as an externally repacked snapshot package, and derive a separate 15-package `snapshotPackageNames` topology.
- Expose a typed snapshot-only repack entry point. It cannot select publish or uncertified modes.
- Make trusted-main validation require DevTools' exact core peers, non-empty Vite peer range, and watcher-only runtime dependencies before attestation or publication.

## Verification

- `node --test .github/scripts/pr-snapshot-artifact.test.mjs`: 13/13 passed, including negative DevTools peer/runtime-boundary cases and generated workflow/topology checks.
- `vitest run scripts/src/commands/devtools-artifact.test.ts`: 15/15 passed for snapshot and release manifest rewrites.
- `devenv tasks run check:quick --no-tui --show-output`: passed after merging current `main`, including TypeScript, formatting, oxlint, Genie, lockfile, source-policy, and megarepo checks.
- Real `release:snapshot:pack:git-sha` proof: packed 15 packages as `0.0.0-snapshot-pr.1447.1d789fbcaaef68b2fef50936410a578e18d164b2`; trusted manifest creation accepted all 15.
- Repacked DevTools tarball SHA-256: `f9ced6987ddbbec23b4c194ce72f80073ea8f9fcd7665e06f34f98885b79b04b`.
- Its manifest contained only `@parcel/watcher@^2.5.0` in runtime dependencies, exact snapshot peers for adapter-web/utils, and the artifact's unchanged `vite@^8.0.16` peer range. Embedded release metadata bound it to build `dt-20260718-51c22feb` with `ci-snapshot` certification.

No package was published during local verification.

## Complexity

The release topology now distinguishes workspace-published packages from externally repacked snapshot packages. The distinction is necessary because DevTools is built in a separate artifact pipeline but must participate in the same immutable npm cohort.

## Concerns

The checked-in artifact currently declares Vite 8 only. This PR deliberately preserves that source contract; a refreshed artifact must widen the source peer range before Vite 7 consumers can install the cohort.

The trusted topology change must merge before any PR head expecting a 15-package candidate can be promoted. Older 14-package candidates remain valid only against the older trusted topology digest.

## Friction & bottlenecks

The real snapshot proof takes about two minutes because it rebuilds and packs the complete release group. No persistent product bottleneck was introduced.

## Follow-ups

- Refresh the checked-in DevTools artifact after its source peer range includes the supported Vite versions.
- Re-run the exact 15-package trusted snapshot proof against that artifact before consuming it downstream.

## References

Closes #1454.

Related: #1458.

No changeset or changelog entry: this changes repository-local release tooling and does not itself publish or alter a package version.

<details>
<summary>Posted on behalf of @schickling</summary>

| field | value |
| --- | --- |
| `agent_name` | ☁️ co3-billow |
| `agent_session_id` | 7f3b9666-034f-4e9e-890b-491957f7f75b |
| `agent_tool` | Codex CLI |
| `agent_tool_version` | 0.144.1 |
| `agent_runtime` | Codex CLI 0.144.1 |
| `agent_model` | unknown |
| `runtime_profile` | /nix/store/1n76sy6i9y3ki3k4w04jksqvygw67sj0-coding-agent-runtime-profile/share/coding-agents/profile.json |
| `skills_manifest` | /nix/store/jn6r0r7r59x3a525jch4pwzwykszqp60-agent-skills-corpus/share/agent-skills/manifest.json |
| `worktree` | livestore-devtools-artifact-peer-deps/schickling-assistant/fix/devtools-artifact-peer-deps |
| `machine` | dev3 |
| `tooling_profile` | dotfiles@4eac376 |
</details>

### CODE REVIEW

File: scripts/src/commands/release.ts [hunk 1]
```diff
@@ -437,6 +442,38 @@ const packPackageForPublish = ({ cwd, pkg, version }: { cwd: string; pkg: string
...
+
+    const fsEffect = yield* FileSystem.FileSystem
+    const workDir = `${cwd}/tmp/release-pack/${version}/livestore-devtools-vite-artifact`
+    yield* fsEffect.remove(workDir, { recursive: true }).pipe(Effect.catch(() => Effect.void))
+    yield* fsEffect.makeDirectory(workDir, { recursive: true })
+
+    const result = yield* Effect.tryPromise({
+      try: () =>
+        repackDevtoolsSnapshotArtifact({
+          manifest: `${cwd}/${descriptor.manifest}`,
```
[the remark this answers is not in the document]
Comment (schickling-assistant) [reply scripts/src/commands/release.ts:467]: Valid and correctly analyzed — thanks. Confirmed the vector is real end-to-end (untrusted PR-head pack → trusted OIDC publish of the PR-packed tarballs; the validator checks package shape but not that the DevTools artifact content is bound to main). One correction: the suggested "validate the digest against main" is not sufficient here — the pack runs PR-provided *code*, so any recorded/embedded digest is attacker-forgeable; a sound fix requires trusted-side *content* verification.

Severity is moderate and gated (same-repo only — forks excluded — plus an approved review on the exact head). The vulnerable external-artifact-repack mechanism is also transitional: `@livestore/devtools-vite` source is moving into `livestore-contrib` to be published source-available, which removes this untrusted-external-tarball bridge entirely. Rather than harden a path that's being retired, tracking the trusted-content-verification hardening in #1483 to land alongside that migration. Resolving here.

**Comment by schickling-assistant** [comment]
Superseded by the DevTools source-available migration — tracked in epic #1497.

The external-artifact machinery this PR extends is retired once `@livestore/devtools-vite` is built from source in `livestore-contrib`. **Holding this rather than landing it** — it's listed for closure as part of #1497's cleanup, after cutover is proven.


<details>
<summary>Posted on behalf of @schickling</summary>

| field | value |
| --- | --- |
| `agent_name` | 🎈 cl1-swell |
| `agent_session_id` | 5a1e2097-a640-400a-a2b8-be55050a58b4 |
| `agent_tool` | Claude Code |
| `agent_tool_version` | 2.1.202 |
| `agent_runtime` | Claude Code 2.1.202 |
| `agent_model` | claude-opus-4-8 |
| `runtime_profile` | /nix/store/393xxjd29yx4b41rsjaglp3r47q9k7s5-coding-agent-runtime-profile/share/coding-agents/profile.json |
| `skills_manifest` | /nix/store/xhnp6r3s5n46kpcq182d56ib1dia1179-agent-skills-corpus/share/agent-skills/manifest.json |
| `worktree` | livestore/schickling/2026-07-19-misc |
| `machine` | dev3 |
| `tooling_profile` | dotfiles@ab9375c |
</details>
